src/linear-types/Control/Distributed/Spark/Safe/Closure.hs

```diff
@@ -0,0 +1,230 @@
+-- | Foreign exports and instances to deal with 'Closure' in Spark.
+
+{-# LANGUAGE QualifiedDo #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StaticPointers #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fplugin=Language.Java.Inline.Plugin #-}
+
+module Control.Distributed.Spark.Safe.Closure
+  ( MapPartitionsFunction(..)
+  , ReduceFunction(..)
+  , ReifyFun(..)
+  , reifyFun_
+  , ReflectFun(..)
+  , JFun1
+  , JFun2
+  ) where
+
+import qualified Prelude 
+import Prelude.Linear as PL
+import Control.Functor.Linear as Linear
+import qualified Data.Functor.Linear as D
+import qualified Unsafe.Linear as Unsafe
+import Data.Unrestricted.Linear ()
+import Control.Monad.IO.Class.Linear
+import qualified System.IO.Linear as LIO
+
+import Control.Distributed.Closure
+import Control.Distributed.Closure.TH
+import Data.Binary (encode, decode)
+import Data.Kind (Type)
+import qualified Data.ByteString.Lazy as LBS
+import Data.ByteString (ByteString)
+import Data.Typeable (Typeable)
+import GHC.TypeLits (Nat)
+import Streaming (Stream, Of)
+
+import Foreign.JNI.Safe as S
+import Foreign.JNI.Types.Safe as S
+import Language.Java.Safe
+import Language.Java.Inline.Safe
+import qualified Language.Java.Unsafe as JUnsafe
+
+pairDict :: Dict c1 -> Dict c2 -> Dict (c1, c2)
+pairDict Dict Dict = Dict
+
+closFun1
+  :: forall a b.
+    Dict (Reify a, Reflect b)
+     -> (a -> b)
+     -> JObjectArray
+  %1 -> LIO.IO JObject
+closFun1 Dict f args = Linear.do
+  (args', a) <- S.getObjectArrayElement args 0
+  Ur a' <- reify_ (unsafeCast a)
+  S.deleteLocalRef args'
+  upcast <$> reflect (f a')
+
+tripleDict :: Dict c1 -> Dict c2 -> Dict c3 -> Dict (c1, c2, c3)
+tripleDict Dict Dict Dict = Dict
+
+closFun2
+  :: forall a b c.
+    Dict (Reify a, Reify b, Reflect c)
+     -> (a -> b -> c)
+     -> JObjectArray
+  %1 -> LIO.IO JObject
+closFun2 Dict f args = Linear.do
+    (args', a) <- second unsafeCast <$> S.getObjectArrayElement args 0
+    (args'', b) <- second unsafeCast <$> S.getObjectArrayElement args' 1
+    Ur a' <- reify_ a
+    Ur b' <- reify_ b
+    S.deleteLocalRef args''
+    upcast <$> reflect (f a' b')
+  where
+    second :: (n %1 -> o) -> (m, n) %1 -> (m, o)
+    second g (a, b) = (a, g b) 
+
+clos2bs :: Typeable a => Closure a %1-> ByteString
```

Comment:
Hmm, that's a good point about `ReifyFun`. I'm a little hesitant to delete it as it could be useful at some point, but then again, I can't really think of any real uses for reifying a java function.